File: js/src/models/GamingActivityEvent.js
/**
 * Gaming Activity Event Model
 * Represents gaming activity events for the Data Pipeline API
 */
class GamingActivityEvent {
  constructor(data = {}) {
    this.account_id = data.account_id;
    this.workspace_id = data.workspace_id;
    this.user_id = data.user_id;
    this.event_category = data.event_category || 'Gaming Activity';
    this.event_name = data.event_name;
    this.event_id = data.event_id;
    this.event_time = data.event_time;
    this.wager_amount = data.wager_amount;
    this.win_amount = data.win_amount;
    this.game_id = data.game_id;
    this.game_title = data.game_title;
    this.provider = data.provider;
    this.game_type = data.game_type;
    this.session_id = data.session_id;
    this.round_id = data.round_id;
    this.device = data.device;
    this.platform = data.platform;
    this.currency = data.currency;
    this.bet_type = data.bet_type;
    this.payout_multiplier = data.payout_multiplier;
    this.loss_amount = data.loss_amount;
    this.bonus_id = data.bonus_id;
    this.free_spin_id = data.free_spin_id;
    this.num_spins_played = data.num_spins_played;
    this.game_theme = data.game_theme;
    this.remaining_spins = data.remaining_spins;
    this.bet_value_per_spin = data.bet_value_per_spin;
    this.wagering_requirements_met = data.wagering_requirements_met;
    this.free_spin_expiry_date = data.free_spin_expiry_date;
    this.campaign_id = data.campaign_id;
    this.campaign_name = data.campaign_name;
    this.rtp = data.rtp;
    this.game_category = data.game_category;
    this.winning_bet_amount = data.winning_bet_amount;
    this.jackpot_type = data.jackpot_type;
    this.volatility = data.volatility;
    this.min_bet = data.min_bet;
    this.max_bet = data.max_bet;
    this.number_of_reels = data.number_of_reels;
    this.number_of_paylines = data.number_of_paylines;
    this.feature_types = data.feature_types;
    this.game_release_date = data.game_release_date;
    this.live_dealer_availability = data.live_dealer_availability;
    this.side_bets_availability = data.side_bets_availability;
    this.multiplayer_option = data.multiplayer_option;
    this.auto_play = data.auto_play;
    this.poker_variant = data.poker_variant;
    this.buy_in_amount = data.buy_in_amount;
    this.table_type = data.table_type;
    this.stakes_level = data.stakes_level;
    this.number_of_players = data.number_of_players;
    this.game_duration = data.game_duration;
    this.hand_volume = data.hand_volume;
    this.player_position = data.player_position;
    this.final_hand = data.final_hand;
    this.rake_contribution = data.rake_contribution;
    this.multi_tabling_indicator = data.multi_tabling_indicator;
    this.session_result = data.session_result;
    this.vip_status = data.vip_status;
    this.blind_level = data.blind_level;
    this.rebuy_and_addon_info = data.rebuy_and_addon_info;
    this.sport_type = data.sport_type;
    this.betting_market = data.betting_market;
    this.odds = data.odds;
    this.live_betting_availability = data.live_betting_availability;
    this.result = data.result;
    this.bet_status = data.bet_status;
    this.betting_channel = data.betting_channel;
    this.bonus_type = data.bonus_type;
    this.bonus_amount = data.bonus_amount;
    this.free_spin_start_date = data.free_spin_start_date;
    this.num_spins_awarded = data.num_spins_awarded;
    this.bonus_code = data.bonus_code;
    this.parent_game_category = data.parent_game_category;
    this.money_type = data.money_type;
    this.transaction_type = data.transaction_type;
  }

  /**
   * Validates the gaming activity event data
   * @returns {Object} Validation result with isValid boolean and errors array
   */
  validate() {
    const errors = [];

    // Required fields
    if (!this.account_id) errors.push('account_id is required');
    if (!this.workspace_id) errors.push('workspace_id is required');
    if (!this.user_id) errors.push('user_id is required');
    if (!this.event_name) errors.push('event_name is required');
    if (!this.event_id) errors.push('event_id is required');
    if (!this.event_time) errors.push('event_time is required');
    if (!this.game_id) errors.push('game_id is required');
    if (!this.game_title) errors.push('game_title is required');

    // Event category validation
    if (this.event_category && this.event_category !== 'Gaming Activity') {
      errors.push('event_category must be "Gaming Activity" for gaming events');
    }

    // Event name validation
    const validEventNames = [
      'Play Casino Game',
      'Game Win',
      'Game Loss',
      'Game Draw',
      'Bonus Game',
      'Free Spins',
      'Tournament Entry',
      'Tournament Win',
      'Progressive Jackpot',
      'Side Bet'
    ];

    if (this.event_name && !validEventNames.includes(this.event_name)) {
      errors.push(`event_name must be one of: ${validEventNames.join(', ')}`);
    }

    // Amount validations
    if (this.wager_amount && (typeof this.wager_amount !== 'number' || this.wager_amount < 0)) {
      errors.push('wager_amount must be a non-negative number');
    }

    if (this.win_amount && (typeof this.win_amount !== 'number' || this.win_amount < 0)) {
      errors.push('win_amount must be a non-negative number');
    }

    // Game type validation
    const validGameTypes = [
      'slots',
      'table_games',
      'card_games',
      'live_casino',
      'bingo',
      'scratch_cards',
      'lottery',
      'sports_betting',
      'virtual_sports'
    ];

    if (this.game_type && !validGameTypes.includes(this.game_type)) {
      errors.push(`game_type must be one of: ${validGameTypes.join(', ')}`);
    }

    // Device validation
    if (this.device && !['desktop', 'mobile', 'tablet', 'app'].includes(this.device)) {
      errors.push('device must be one of: desktop, mobile, tablet, app');
    }

    // Platform validation
    const validPlatforms = [
      'web',
      'ios',
      'android',
      'windows',
      'mac',
      'linux'
    ];

    if (this.platform && !validPlatforms.includes(this.platform)) {
      errors.push(`platform must be one of: ${validPlatforms.join(', ')}`);
    }

    // Currency validation
    if (this.currency && !this.isValidCurrency(this.currency)) {
      errors.push('currency must be a valid 3-letter ISO currency code');
    }

    // Date format validation
    if (this.event_time && !this.isValidDateTime(this.event_time)) {
      errors.push('event_time must be in ISO 8601 format (YYYY-MM-DDTHH:mm:ssZ)');
    }

    return {
      isValid: errors.length === 0,
      errors
    };
  }

  /**
   * Validates ISO 8601 datetime format
   * @param {string} dateTime - DateTime string to validate
   * @returns {boolean} True if datetime is valid
   */
  isValidDateTime(dateTime) {
    const date = new Date(dateTime);
    return date instanceof Date && !isNaN(date);
  }

  /**
   * Validates currency code format
   * @param {string} currency - Currency code to validate
   * @returns {boolean} True if currency code is valid
   */
  isValidCurrency(currency) {
    const currencyRegex = /^[A-Z]{3}$/;
    return currencyRegex.test(currency);
  }

  /**
   * Converts the model to a plain object
   * @returns {Object} Plain object representation
   */
  toJSON() {
    const obj = {};
    Object.keys(this).forEach(key => {
      if (this[key] !== undefined) {
        obj[key] = this[key];
      }
    });
    return obj;
  }

  /**
   * Creates a GamingActivityEvent instance from plain object
   * @param {Object} data - Plain object data
   * @returns {GamingActivityEvent} New GamingActivityEvent instance
   */
  static fromObject(data) {
    return new GamingActivityEvent(data);
  }
}

module.exports = GamingActivityEvent;

File: python/src/python/models/GamingActivityEvent.py
from dataclasses import dataclass, field, asdict
from typing import Optional, Dict, Any, List
import re
from datetime import datetime


@dataclass
class GamingActivityEvent:
    account_id: Optional[str] = None
    workspace_id: Optional[str] = None
    user_id: Optional[str] = None
    event_category: str = "Gaming Activity"
    event_name: Optional[str] = None
    event_id: Optional[str] = None
    event_time: Optional[str] = None
    wager_amount: Optional[float] = None
    win_amount: Optional[float] = None
    game_id: Optional[str] = None
    game_title: Optional[str] = None
    provider: Optional[str] = None
    currency: Optional[str] = None
    loss_amount: Optional[float] = None
    bonus_id: Optional[str] = None
    free_spin_id: Optional[str] = None
    jackpot_amount: Optional[float] = None
    num_spins_played: Optional[int] = None
    game_theme: Optional[str] = None
    remaining_spins: Optional[int] = None
    bet_value_per_spin: Optional[float] = None
    wagering_requirements_met: Optional[bool] = None
    free_spin_expiry_date: Optional[str] = None
    campaign_id: Optional[str] = None
    campaign_name: Optional[str] = None
    rtp: Optional[float] = None
    game_category: Optional[str] = None
    winning_bet_amount: Optional[float] = None
    jackpot_type: Optional[str] = None
    volatility: Optional[str] = None
    min_bet: Optional[float] = None
    max_bet: Optional[float] = None
    number_of_reels: Optional[int] = None
    number_of_paylines: Optional[int] = None
    feature_types: Optional[str] = None
    game_release_date: Optional[str] = None
    live_dealer_availability: Optional[bool] = None
    side_bets_availability: Optional[bool] = None
    multiplayer_option: Optional[bool] = None
    auto_play: Optional[bool] = None
    poker_variant: Optional[str] = None
    tournament_name: Optional[str] = None
    buy_in_amount: Optional[float] = None
    table_type: Optional[str] = None
    stakes_level: Optional[str] = None
    number_of_players: Optional[int] = None
    game_duration: Optional[int] = None
    hand_volume: Optional[int] = None
    player_position: Optional[str] = None
    final_hand: Optional[str] = None
    rake_contribution: Optional[float] = None
    multi_tabling_indicator: Optional[bool] = None
    session_result: Optional[str] = None
    vip_status: Optional[str] = None
    blind_level: Optional[str] = None
    rebuy_and_addon_info: Optional[str] = None
    sport_type: Optional[str] = None
    betting_market: Optional[str] = None
    odds: Optional[float] = None
    live_betting_availability: Optional[bool] = None
    result: Optional[str] = None
    bet_status: Optional[str] = None
    betting_channel: Optional[str] = None
    bonus_type: Optional[str] = None
    bonus_amount: Optional[float] = None
    free_spin_start_date: Optional[str] = None
    num_spins_awarded: Optional[int] = None
    bonus_code: Optional[str] = None
    parent_game_category: Optional[str] = None
    money_type: Optional[str] = None
    transaction_type: Optional[str] = None

    # -------------------------------
    # Validation methods
    # -------------------------------
    def validate(self) -> Dict[str, Any]:
        errors: List[str] = []

        # Required fields
        if not self.account_id:
            errors.append("account_id is required")
        if not self.workspace_id:
            errors.append("workspace_id is required")
        if not self.user_id:
            errors.append("user_id is required")
        if not self.event_name:
            errors.append("event_name is required")
        if not self.event_id:
            errors.append("event_id is required")
        if not self.event_time:
            errors.append("event_time is required")
        if not self.game_id:
            errors.append("game_id is required")
        if not self.game_title:
            errors.append("game_title is required")

        # Event category validation
        if self.event_category and self.event_category != "Gaming Activity":
            errors.append('event_category must be "Gaming Activity" for gaming events')

        # Event name validation
        valid_event_names = [
            "Play Casino Game",
            "Game Win",
            "Game Loss",
            "Game Draw",
            "Bonus Game",
            "Free Spins",
            "Tournament Entry",
            "Tournament Win",
            "Progressive Jackpot",
            "Side Bet"
        ]
        if self.event_name and self.event_name not in valid_event_names:
            errors.append(f"event_name must be one of: {', '.join(valid_event_names)}")

        # Amount validations
        if self.wager_amount is not None and (self.wager_amount < 0):
            errors.append("wager_amount must be a non-negative number")
        if self.win_amount is not None and (self.win_amount < 0):
            errors.append("win_amount must be a non-negative number")

        # Currency validation
        if self.currency and not self.is_valid_currency(self.currency):
            errors.append("currency must be a valid 3-letter ISO currency code")

        # Date validation
        if self.event_time and not self.is_valid_datetime(self.event_time):
            errors.append("event_time must be in ISO 8601 format (YYYY-MM-DDTHH:mm:ssZ)")

        return {"isValid": len(errors) == 0, "errors": errors}

    @staticmethod
    def is_valid_datetime(date_time: str) -> bool:
        try:
            datetime.fromisoformat(date_time.replace("Z", "+00:00"))
            return True
        except ValueError:
            return False

    @staticmethod
    def is_valid_currency(currency: str) -> bool:
        return re.match(r"^[A-Z]{3}$", currency) is not None

    # -------------------------------
    # Utility methods
    # -------------------------------
    def to_dict(self) -> Dict[str, Any]:
        return {k: v for k, v in asdict(self).items() if v is not None}

    @classmethod
    def from_object(cls, data: Dict[str, Any]) -> "GamingActivityEvent":
        return cls(**data)

    def __repr__(self) -> str:
        return f"GamingActivityEvent(event_id={self.event_id}, event_name={self.event_name}, user_id={self.user_id})"
